src/Microsoft.AspNetCore.Blazor.Razor.Extensions/BlazorNodeWriter.cs

```diff
@@ -49,5 +53,169 @@ public sealed override void WriteCSharpCodeAttributeValue(CodeRenderingContext c
             context.Diagnostics.Add(BlazorDiagnosticFactory.Create_CodeBlockInAttribute(node.Source, content));
             return;
         }
+
+
+        // Currently the same for design time and runtime
+        public void WriteComponentTypeInferenceMethod(CodeRenderingContext context, ComponentTypeInferenceMethodIntermediateNode node)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            // This is ugly because CodeWriter doesn't allow us to erase, but we need to comma-delimit. So we have to
+            // materizalize something can iterate, or use string.Join. We'll need this multiple times, so materializing
+            // it.
+            var parameters = GetParameterDeclarations();
+
+            // This is really similar to the code in WriteComponentAttribute and WriteComponentChildContent - except simpler because
+            // attributes and child contents look like variables. 
+            //
+            // Looks like:
+            //
+            //  public static void CreateFoo_0<T1, T2>(RenderTreeBuilder builder, int seq, int __seq0, T1 __arg0, int __seq1, global::System.Collections.Generic.List<T2> __arg1, int __seq2, string __arg2)
+            //  {
+            //      builder.OpenComponent<Foo<T1, T2>>();
+            //      builder.AddAttribute(__seq0, "Attr0", __arg0);
+            //      builder.AddAttribute(__seq1, "Attr1", __arg1);
+            //      builder.AddAttribute(__seq2, "Attr2", __arg2);
```

Comment:
Ah yes. I think last time we spoke, I suggested we could do away with sequence numbers for attribute nodes, since the diffing is meant to be order-independent anyway. But on futher inspection, I now remember we have two different merge strategies for attributes: a fast path that does use sequence numbers to avoid needing to remember what it saw earlier, and a slow path that's independent of sequence numbers.

So I agree it's better to preserve the sequence numbers like you've done here.

Reply:
Yeah, it seemed worth it to me regardless of whether it had an actual runtime impact. This way the behaviour of tests that render a component is the same with and without type inference.